### DISCUSSION TITLE
Plane Stress assumption in Chaboche Viscoplasticity

### DISCUSSION THREAD

**Discussion by VaishnavBhaskaran** [thelfer/tfel#601] [Q&A] [answered]
Hello!
I am using ChabocheViscoplasticity material behaviour to model a simple plate. I have two issues in implementing Plane Stress using this library.

Issue 1:
I am using legacy FEniCS. I am using an older version of Chaboche (please refer Old_ChabocheViscoplasticity.mfront). Everything works fine with this if Plane Strain is assumed but it doesn't have Plane Stress behaviour in it.  Is there a way to include Plane Stress in the same code?

![image](https://github.com/user-attachments/assets/b26666f6-42f2-4ff4-a623-a25cabb84680)

Issue 2:
I downloaded the latest version of Chaboche (please refer New_ChabocheViscoplasticity.mfront) which has the following libraries

![image](https://github.com/user-attachments/assets/98b1a482-c6e1-47d3-bb49-3c7309b25f23)

If I use this in my code (with legacy FEniCS), I get the following error after some increments for Plane Strain.

![image](https://github.com/user-attachments/assets/1db490c6-dfd6-4a63-bb72-393e25391bba)

I change it to Plane Stress and still it cannot load the behaviour

`Traceback (most recent call last):
  File "/home/vaishnav27/Fenics_Simulation/test.py", line 59, in <module>
    b = mgis_bv.load('/home/vaishnav27/Fenics_Simulation/New_folder/src/libBehaviour.so','ChabocheViscoplasticity',h)
RuntimeError: LibrariesManager::getBehaviour: can't load behaviour 'ChabocheViscoplasticity' in library '/home/vaishnav27/Fenics_Simulation/New_folder/src/libBehaviour.so' for hypothesis 'PlaneStress'`

Is FEniCS version causing this issue or is it something else?

It's been only a few weeks since I started using FEniCS and mfront. I apologize if I have missed anything fundamental.
[New folder.zip](https://github.com/user-attachments/files/16244922/New.folder.zip)

**Comment by thelfer** [comment, marked as answer]
Hi @VaishnavBhaskaran,
I am currently on vacation, I can't read your zip file on my phone. Can you copy/paste the Mfront implémentations ?
However, support for plane stress must bé explicitly requêtes using:

@ModellingHypotheses ".+";

But this requires that thé behaviour Can Indeed support plane stress

**Reply by VaishnavBhaskaran** [reply]
Hello @thelfer,
I am calling the behaviour from my code with `mgis.behaviour.Hypothesis.PlaneStress`

### This is the latest Mfront code:

````
@DSL       Implicit;
Author    Jean-Michel Proix;
Date      15/04/2014;
Behaviour ChabocheViscoplasticity;
Brick     StandardElasticity;

Theta 1.;
Epsilon 1.e-13;

@MaterialProperty stress young;
young.setGlossaryName("YoungModulus");
@MaterialProperty real nu;
nu.setGlossaryName("PoissonRatio");

@StateVariable strain p;
p.setGlossaryName("EquivalentViscoplasticStrain");
@StateVariable StrainStensor a[2];

@MaterialProperty stress Rinf,R0;
@MaterialProperty real b;
@MaterialProperty stress C[2];
@MaterialProperty real   g[2];
@MaterialProperty real m;
@MaterialProperty real K;

LocalVariable bool bloading;

@InitLocalVariables{
  StressStensor scin = computeElasticPrediction();
  for(unsigned short i=0;i!=2;++i){
     scin  -= 2*C[i]*a[i]/3;
  }
  const stress seq = sigmaeq(scin);
  const stress Rp  = Rinf + (R0-Rinf)*exp(-b*p);  
  bloading = seq - Rp > 0;
}

Integrator{
  constexpr const real eps = 1.e-12;
  if(bloading){
    const auto mu      = computeMu(young,nu);
    const strain p_    = p +theta*dp ;
    StressStensor scin = sig;
    StrainStensor a_[2];
    for(unsigned short i=0;i!=2;++i){
      a_[i]  = a[i]+theta*da[i];
      scin  -= 2*C[i]*a_[i]/3;
    }
    const stress seq = sigmaeq(scin);
    const stress Rp  = Rinf + (R0-Rinf)*exp(-b*p_) ;  
    const stress F = max(seq - Rp,stress(0));
    const real Fexp   = pow(F/K,m-1)/K;
    const auto iseq = 1/max(seq,eps*young);
    const Stensor n = 3*deviator(scin)*iseq/2;
    const auto vp  = Fexp*F;
    feel += dp*n;
    fp   -= vp*dt;
    for(unsigned short i=0;i!=2;++i){
      fa[i] = da[i] - dp*(n-g[i]*a_[i]);
    }
    const Stensor4 Jmn  = Stensor4::M() - (n^n);    
    dfeel_ddeel += 2*mu*theta*dp*Jmn*iseq ;
    dfeel_ddp    = n;
    dfp_ddeel    = - 2*Fexp*m*dt* 2.*mu*theta*(n|Stensor4::M())/3;
    dfp_ddp      = 1 + theta* Fexp *m *dt*b*(Rinf-Rp);
    for(unsigned short i=0;i!=2;++i){
      dfeel_dda(i)   = -2*C[i]*dp*theta* iseq/3*Jmn ;
      dfp_dda(i)     =  2*Fexp*m*dt*C[i]*theta/3*n;  
      dfa_ddeel(i)   = -2*mu*theta*dp*Jmn *iseq;
      dfa_ddp(i)     = -n + g[i]*a_[i];
      dfa_dda(i,i)   =  (1+dp*g[i]*theta)*Stensor4::Id()+C[i]*dp*theta*iseq/1.5*Jmn;
    }
    dfa_dda(0,1)   =  2*C[1]*dp*theta*iseq/3*Jmn;
    dfa_dda(1,0)   =  2*C[0]*dp*theta*iseq/3*Jmn;
  }
}
````



### This is the older version:

````
@DSL       ImplicitParser;
Behaviour ChabocheViscoplasticity;
Brick     StandardElasticity;
Algorithm NewtonRaphson;
Epsilon 1.e-7;

@MaterialProperty stress young;
young.setGlossaryName("YoungModulus");
@MaterialProperty real nu;
nu.setGlossaryName("PoissonRatio");

@StateVariable strain p;
p.setGlossaryName("EquivalentViscoplasticStrain");
@StateVariable StrainStensor a[2];

@MaterialProperty stress Rinf,R0;
@MaterialProperty real b;
@MaterialProperty stress C[2];
@MaterialProperty real   g[2];
@MaterialProperty real m;
@MaterialProperty real K;

LocalVariable stress mu;

/* Initialize Lame coefficients */
@InitializeLocalVariables{
  mu = computeMu(young,nu);
} 

Integrator{
  const real two_third     = 2/real(3);
  constexpr const real eps = real(1.e-7);
  const auto p_ = p +theta*dp ;
  auto scin     = sig ;
  StrainStensor a_[2];
  for(unsigned short i=0;i!=2;++i){
     a_[i]  = a[i]+theta*da[i];
     scin  -= C[i]*a_[i]*two_third;
  }
  const auto seq = sigmaeq(scin);
  const auto Rp  = Rinf + (R0-Rinf)*exp(-b*p_) ;  
  const auto F   = seq - Rp ;
  if(F > 0){
    const auto Fexp = pow(F/K,m-1)/K ;
    const auto iseq = 1/(max(seq,eps*young));
    const auto n    = 3*iseq*deviator(scin)/2;
    const auto vp   = Fexp*F;  
    feel += dp*n;
    fp   -= vp*dt;
    for(unsigned short i=0;i!=2;++i){
      fa[i] -= dp*(n-g[i]*a_[i]);
    }
    // computation of the jacobian matrix
    const auto Jmn = eval(Stensor4::M()-(n^n));    
    dfeel_ddeel += 2*mu*theta*dp*Jmn*iseq ;
    dfeel_ddp    = n;
    dfp_ddeel    = - Fexp*m*dt* 2*mu*theta*(n| Stensor4::M() )*two_third;
    dfp_ddp     += theta* Fexp *m *dt*b*(Rinf-Rp);
    for(unsigned short i=0;i!=2;++i){
      dfeel_dda(i)   = -C[i]*dp*theta*iseq*two_third*Jmn ;
      dfp_dda(i)     =  Fexp*m*dt*C[i]*theta*two_third*n;  
      dfa_ddeel(i)   = -2*mu*theta*dp*Jmn *iseq;
      dfa_ddp(i)     = -n + g[i]*a_[i];
      dfa_dda(i,i)   =  (1+dp*g[i]*theta)*Stensor4::Id()+C[i]*dp*theta*iseq*two_third*Jmn;
    }
    dfa_dda(0,1)   =  C[1]*dp*theta*iseq*two_third*Jmn;
    dfa_dda(1,0)   =  C[0]*dp*theta*iseq*two_third*Jmn;
  }
}
````

**Reply by Algorithm** [reply]
You are copying in @algorithm. We're therefore getting all of your responses. You may want to remove reference to our username.

**Reply by maxence-wz** [reply]
Hi @VaishnavBhaskaran, 

you need to use the ```@ModellingHypotheses {".+"};``` option to activate the support of plane stress. 

```
@DSL       Implicit;
@Author    Jean-Michel Proix;
@Date      15/04/2014;
@Behaviour ChabocheViscoplasticity;
@Brick     StandardElasticity;

@ModellingHypotheses {".+"};

...

```

**Reply by thelfer** [reply]
The older version uses a very loose convergence threshold. Would you try to use the same value and sée if you have similar behaviours ?

**Reply by VaishnavBhaskaran** [reply]
Hello  @thelfer,
I tried with the same value and still get the same error.

Hello @maxence-wz,
I added `@ModellingHypotheses {".+"};` option in the latest Mfront code but I am still not able to call Plane Stress Hypotheses.

![image](https://github.com/user-attachments/assets/704e207e-70a0-4249-9b78-44bbb736cfca)

I am calling it in my code like this 

```python
h = mgis_bv.Hypothesis.PlaneStress
b = mgis_bv.load('/home/vaishnav27/Fenics_Simulation/New_folder/src/libBehaviour.so','ChabocheViscoplasticity',h)
m = mgis_bv.MaterialDataManager(b, ngauss)
```

**Reply by VaishnavBhaskaran** [reply]
Thank you @maxence-wz! Compiling the MFront file again after adding `@ModellingHypotheses {".+"};` to it, solves the issue of using Plane Stress option. It works for both the old and new versions of the code.

There is still the issue of convergence with the newer version of the code. I'll check it with DOLFINx to see if the legacy FEniCS is causing this issue.

**Reply by maxence-wz** [reply]
@VaishnavBhaskaran both files work for me (with the MTest application), at least with a plane strain hypothesis. I don't have any convergence problems.

In addition, an example of Chaboche's law implemented in MFront format for plane stress is available [here](https://github.com/ralic/Code_Aster/blob/master/astest/Chaboche.mfront).

**Reply by maxence-wz** [reply]
@VaishnavBhaskaran, as you use the ```StandardElasticity``` brick, plane stress is automatically handled by the generated law ([see](https://thelfer.github.io/tfel/web/BehaviourBricks.html)). So no need for the changes in the example [here](https://github.com/ralic/Code_Aster/blob/master/astest/Chaboche.mfront), sorry for the confusion...

**Reply by VaishnavBhaskaran** [reply]
No problem @maxence-wz!

I tried replacing the `IsotropicLinearHardeningPlasticity.mfront` from this [tutorial](https://comet-fenics.readthedocs.io/en/latest/demo/plasticity_mfront/plasticity_mfront.py.html#plasticitymfront) with both the versions of `ChabocheViscoplasticity.mfront` but still the newer version resulted in the same convergence error. The older version was running fine.

**Reply by maxence-wz** [reply]
@VaishnavBhaskaran the convergence error only occurs for plane stress hypothesis?

**Reply by VaishnavBhaskaran** [reply]
No, I tried with Plane Strain. I don't think the convergence is related to the Plane Stress hypothesis

**Reply by maxence-wz** [reply]
@VaishnavBhaskaran are you sure of your material property values in your test file? In particular, the value of m.

**Comment by thelfer** [comment]
Hi @VaishnavBhaskaran,
I checked the old and new versions. The new one seems much better IMHO, although quite similar to the old one. Both shall work as expected, but you do have to use a very low convergence criterion (typically 1e-13 is fine).
I need to investigate further.
Thomas

**Reply by thelfer** [reply]
As stated by @maxence-wz, maybe the values of the material properties must be checked.

**Reply by thelfer** [reply]
To do this, would you try to make a simple MTest file for an unaxial test (cyclic) ?

**Reply by VaishnavBhaskaran** [reply]
Hello @thelfer, @maxence-wz,
I ran a sample simulation with uniaxial loading of a plate with the older version and the results look fine

![image](https://github.com/user-attachments/assets/930f1aed-4f9a-4658-84ba-fac56bb1c6c2)

So I guess the material parameters are fine. I guess I'll stick with the older version for now as I have some time constraints. But as stated, it shouldn't bring much difference to the results.

Thank you for your time and effort!

**Reply by thelfer** [reply]
Would you upload your input files, if different from your initial post ?

**Reply by VaishnavBhaskaran** [reply]
Sure. I am attaching the [code](https://github.com/user-attachments/files/16468670/Test_Simulation.zip) for the above result. The MFront files are same as in the initial post with only the addition of `@ModellingHypotheses ".+";` for Plane Stress condition.

**Reply by VaishnavBhaskaran** [reply]
Hello again!

I was plotting the stress-strain graph and find that the behaviour is linear irrespective of the load as shown below

![image](https://github.com/user-attachments/assets/5dd780e5-ce5a-4479-bdc8-ae3901b1e846)

Do I have to do anything apart from these? The update of constitutive law is handled by the MFrontNonlinearProblem class in the background as I understand from the source codes

```python
u = Function(V, name = "displacement")
material = mf.MFrontNonlinearMaterial("./src/libBehaviour.so", "ChabocheViscoplasticity",
                                      hypothesis = hypothesis,
                                      material_properties = mat_prop)
problem = mf.MFrontNonlinearProblem(u, material, quadrature_degree = 2, bcs = bcs)
problem.set_loading(load*dot(n, u)*ds(RIGHT_EDGE))

for i, t in enumerate(load_steps):
    load.t = t
    problem.solve(u.vector())
    p0 = problem.get_state_variable("EquivalentViscoplasticStrain", project_on = ("DG", 0))
    p1 = problem.get_flux("Stress", project_on = ("DG", 0))
    epsilon = problem.get_gradient("Strain", project_on=("DG", 0))
``` 

Please find the full implementation [here](https://github.com/user-attachments/files/16548734/Fenics_Simulation.zip).

**Reply by thelfer** [reply]
The yield strength is egal to 120MPa, a value not reached on the output. Is it normal ?

**Reply by Vaishnav276** [reply]
Hello,
I tried increasing the load but I still get a linear graph. I took another set of material properties to check if they were causing this issue bit I still get the same linear graph as below (Stress vs Strain %)

![Figure_1](https://github.com/user-attachments/assets/31ff1555-4566-475e-b418-126bed65d7ef)

There seems to be no viscoplastic strain at all even though the displacement is large as shown below

![image](https://github.com/user-attachments/assets/f8174b86-2039-4cfb-b90f-43ddb00d2467)

![image](https://github.com/user-attachments/assets/af610f44-4a74-4dcd-a0af-56a15fe5f2fb)


I would really appreciate any suggestion to solve this as I'm stuck with this for a couple of weeks now. I am attaching the code and the xdmf file [here](https://github.com/user-attachments/files/16628029/FS_Small_Strain_Plasticity.zip) for reference.

**Reply by thelfer** [reply]
I got a clue: the time increment is not defined. You can check this in your MFront file by printing the `dt` variable as follows:

~~~~c++
@InitLocalVariables{
...
std::cout << "dt: " << dt << std::endl;
}
~~~~

As a consequence, the behaviour is indeed elastic (the inelastic strain increment is proportional to `dt`).

(Your example work for a plastic behaviour where `dt` is not needed).

Apparently, you can set `dt` as follows:

~~~~python
problem.dt = dt
~~~~

in your `FEniCS` script

**Reply by VaishnavBhaskaran** [reply]
You are absolutely right! I defined the time increment as follows:
```python
t_old = 0.0
for (i, t) in enumerate(load_steps):
    dt = (t - t_old)/100
    loading.t = t
    problem.dt = dt
    problem.solve(u.vector())
    ...
    t_old = t    ### At the end of increment
``` 
Now the curve is nonlinear as shown below

![image](https://github.com/user-attachments/assets/73484bf9-dfb4-42d7-9e45-f3987023d2fa)

But if I increase the load further the solver doesn't converge. I adjusted the tolerance but still doesn't converge for higher loads. Also the EquivalentViscoplasticStrain is still zero.

I tried the same problem setup with `IsotropicLinearHardeningPlasticity` behaviour and I am able to get a good curve as below

![image](https://github.com/user-attachments/assets/e256bbe6-7a13-4e3c-941d-9e33c3514486)

So I'm guessing the problem is with the material properties or there's something still missing. I took the current set of values from this [paper](https://www.sciencedirect.com/science/article/pii/0749641989900156?fr=RR-2&ref=pdf_download&rr=8b4007223d79bb4d) and Mechanics of Solid Materials by J.L. Chaboche.

**Reply by thelfer** [reply]
Have you plot the stresses just before the divergence of the computations ?

**Reply by VaishnavBhaskaran** [reply]
Hello,

everything works fine. I am able to get the full plot as shown below
![Chaboche_Stress_Strain_458](https://github.com/user-attachments/assets/61de2620-cfec-4a85-93d4-7fce9b9801dc)

I was just plotting the results for the wrong point. The above graph is for a point closer to the region where maximum plastic deformation occurs. Increasing the load further leads to failure hence the solver doesn't converge.

**Reply by thelfer** [reply]
So everything is ok, right ?

**Reply by VaishnavBhaskaran** [reply]
Yes. Thank you very much for your timely help!
Just a small question. Are there any tutorials for implementing a dynamic analysis with MFront?

**Reply by thelfer** [reply]
Dynamic aspects are handled by the solver. From the behaviour point of view, nothing changes...

P.S.: This is not exaclty true: the behaviour can provide additional information to help the solver, but it's quite advanced

**Reply by thelfer** [reply]
BTW @bleyerj has published some nice tutorial to make dynamic computations with FEniCs (probably FeniCsX)

**Reply by VaishnavBhaskaran** [reply]
Okay. Thank you so much! I'll have a look at those.
